content/blog/2022/2022-01-25-annotation-processing.md

```diff
@@ -0,0 +1,951 @@
+
+    private int id;
+    private String name;
+
+    public SingleValueAnnotatedEmployee(int id, String name) {
+        this.id = id;
+        this.name = name;
+    }
+
+    public void getEmployeeDetails(){
+
+        System.out.println("Employee Id: " + id);
+        System.out.println("Employee Name: " + name);
+    }
+}
+```
+
+Run a test as below:
+
+```
+public class TestSingleValueAnnotatedEmployee {
+
+    public static void main(String[] args) {
+
+        SingleValueAnnotatedEmployee employee = new SingleValueAnnotatedEmployee(1, "John Doe");
+        employee.getEmployeeDetails();
+
+        Annotation companyAnnotation = employee.getClass().getAnnotation(SingleValueAnnotationCompany.class);
+        SingleValueAnnotationCompany company = (SingleValueAnnotationCompany)companyAnnotation;
+
+        System.out.println("Company Name: " + company.value());
+    }
+}
+```
+
+The single value 'XYZ' overrides the default annotation value and the output is as below:
+
+```
+Employee Id: 1
+Employee Name: John Doe
+Company Name: XYZ
+```
+
+### Full Annotations
+
+They consist of multiple name value pairs. For eg. Company(name="ABC", city="XYZ"). Considering our original Company example:
+
+```java
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Company{
+    String name() default "ABC";
+    String city() default "XYZ";
+}
+```
+
+Create the MultiValueAnnotatedEmployee class as below. Specify the parameters and values as below. The default values will be over-written.
+
+```java
+@Company(name = "AAA", city = "ZZZ")
+public class MultiValueAnnotatedEmployee {
+    
+}
+```
+
+Run a test as below:
+
+```java
+public class TestMultiValueAnnotatedEmployee {
+
+    public static void main(String[] args) {
+
+        MultiValueAnnotatedEmployee employee = new MultiValueAnnotatedEmployee();
+
+        Annotation companyAnnotation = employee.getClass().getAnnotation(Company.class);
+        Company company = (Company)companyAnnotation;
+
+        System.out.println("Company Name: " + company.name());
+        System.out.println("Company City: " + company.city());
+    }
+}
+```
+
+The output is as below, and has overridden the default annotation values:
+
+```
+Company Name: AAA
+Company City: ZZZ
+```
+
+
+
+## A real-world example
+
+For our real-world example, we are going to do a simple simulation of the annotation @Test in JUnit. By marking our functions with the @Test annotation we can determine at run-time, the methods in a class that need to be run as tests. 
+
+We first create the Annotation as below:
+
+```java
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD) 
+public @interface Test {
+}
+```
+
+Next, we create a class AnnotatedMethods, to which we will apply the @Test annotations to some of its methods:
+
+```java
+public class AnnotatedMethods {
+
+    @Test
+    public void Test_1() {
+
+        System.out.println("This is the first test");
+    }
+
+    public void Test_2() {
```

Comment:
Please explain in the text why you left the `@Test` annotation away here. I guess it's to show that it will only work for the first method? If you don't explain it, it will look like an error.